python/phase-1.md

```diff
@@ -0,0 +1,441 @@
+* **NOTE: These will have proper docs and types when developed!**
+
+### Core
+
+Exported from `temporalio` (but probably in `temporalio/core/core.py`, exported from `temporalio.core` then
+exported from `temporalio`):
+
+```python
+class Core:
+    @staticmethod
+    async def create(
+        addr: str,
+        *,
+        namespace: str = "default"
+        # ... additional options omitted for brevity
+    ) -> Core:
+        pass
+
+    @classmethod
+    async def default_local(cls) -> Core:
+        pass
+
+    async def shutdown(self) -> None:
+        pass
+```
+
+Notes:
+
+* After discussion, it has been decided that both clients and workers will be backed by core
+    * Yes that means clients require the Rust library which we accept
+* After discussion, it has been decided that we will not have an implicit local default like TypeScript SDK, but instead
+  an explicit default
+  * `default_local()` memoizes success
+* Should we have a constructor, an `__aenter__`, and an `__aexit__`?
+  * Probably
+
+### Client
+
+Exported from `temporalio` (but probably in `temporalio/client/client.py`, exported from `temporalio.client` then
+exported from `temporalio`):
+
+```python
+class Client:
+    def __init__(core_or_service: temporalio.Core) -> None:
+        pass
+
+    # Raw gRPC service
+    @property
+    def service(self) -> workflowservice.WorkflowService:
+        pass
+
+    async def start_workflow(
+        self,
+        workflow: str,
+        *args: Any,
+        task_queue: str,
+        id: Optional[str] = None,
+        retry_policy: Optional[temporalio.RetryPolicy] = None,
+        # Note, we are embedding signal-with-start in here
+        start_signal: Optional[str] = None,
+        start_signal_args: list[Any] = [],
+        # ... additional options omitted for brevity
+    ) -> temporalio.WorkflowHandle[Any]:
+        pass
+
+    # In a future phase maybe:
+    # @overload
+    # async def start_workflow(workflow: Callable[P, T], *args: P.args, ...) -> temporalio.WorkflowHandle[T]
+    #
+    # @overload
+    # async def start_workflow(workflow: T, *args) -> T
+
+    # Just a wrapper for start workflow that awaits the handle
+    async def execute_workflow(
+        self,
+        workflow: str,
+        *args: Any,
+        # ... additional options omitted for brevity
+    ) -> Any:
+        return await (await self.start_workflow(workflow, *args))
+
+    # Does not validate whether a workflow exists
+    def get_workflow_handle(
+        workflow_id: str,
+        run_id: Optional[str] = None
+        # ...
+    ) -> temporalio.WorkflowHandle[Any]:
+        pass
+
+
+class WorkflowHandle(Generic[T]):
+    def __init__(
+        self,
+        client: temporalio.Client,
+        workflow_id: str,
+        *,
+        run_id: Optional[str] = None
+        # ...
+    ) -> None:
+        pass
+
+    def __await__(self) -> T:
+        pass
```

Comment:
Another issue with this approach is that `await client.get_workflow_handle()` now waits will wait for completion which could be confusing.

Reply:
Let's reconsider adding an explicit `result` method.

Reply:
> Another issue with this approach is that await client.get_workflow_handle() now waits will wait for completion which could be confusing.

This is only unclear because I was forced to change the name. `await client.workflow("my workflow")` was very clear originally from the previous PR.

Reply:
I don't think the name change is what's making it unclear.
It's that you get 2 completely different results with an without `await`.

Reply:
That's normal I think for awaitable objects that carry other info. I will change to adding a result method though.

Reply:
Added `result()`